# Comparação entre medidas estatisticas feitas com numpy e lista nativa do python

In [12]:
from DataFruta_V1.ListaSalarios import ListaSalarios
import numpy as np

tamanho = 1000

salarios = ListaSalarios.geraListaSalarios(tamanho)

lista_salarios = salarios.lista

## Médias e medidas de valor central

### Média aritmética

In [13]:
# usando sem metodos numpy
%timeit salarios.mediaAritmetica()
print("Media dos salario = " + str(salarios.mediaAritmetica()))

103 µs ± 6.38 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Media dos salario = 5700.137827403773


In [14]:
# usando metodos numpy
%timeit np.mean(lista_salarios)
print("Media dos salario = " + str(np.mean(lista_salarios)))

7.64 µs ± 389 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
Media dos salario = 5700.137827403758


### Media geométrica

In [15]:
# usando sem metodos numpy
%timeit salarios.mediaGeometrica()
print("Media geometrica dos salario = " + str(salarios.mediaGeometrica()))

c:\Users\evera\Documents\workspace\Trilha_Python\DataFruta_V1\ListaSalarios.py:125: RuntimeWarning: overflow encountered in scalar multiply
  produto *= i


142 µs ± 16.7 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Media geometrica dos salario = inf


In [16]:
# usando metodos numpy
%timeit np.prod(lista_salarios)**(1/len(lista_salarios))
print("Media geometrica dos salario = " + str(np.prod(lista_salarios)**(1/len(lista_salarios))))

12.2 µs ± 1.49 µs per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
Media geometrica dos salario = inf


### Media harmonica

In [17]:
# usando sem metodos numpy
%timeit salarios.mediaHarmonica()
print("Media harmonica dos salario = " + str(salarios.mediaHarmonica()))

207 µs ± 12.2 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Media harmonica dos salario = 4310.529895667538


In [18]:
# usando metodos numpy
%timeit len(lista_salarios) / np.sum(1/np.array(lista_salarios))
print("Media harmonica dos salario = " + str(len(lista_salarios) / np.sum(1/np.array(lista_salarios))))

10.9 µs ± 569 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
Media harmonica dos salario = 4310.52989566754


### Mediana

In [37]:
# usando sem metodos numpy
%timeit salarios.mediana()
print("Mediana dos salario = " + str(salarios.mediana()))

16.4 µs ± 820 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
Mediana dos salario = 5710.341638400105


In [38]:
# usando metodos numpy
%timeit np.median(lista_salarios)
print("Mediana dos salario = " + str(np.median(lista_salarios)))

33.6 µs ± 2.59 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Mediana dos salario = 5710.341638400105


### Mediana inferior

In [22]:
# usando sem metodos numpy
%timeit salarios.medianaInferior()
print("Mediana inferior dos salario = " + str(salarios.medianaInferior()))

269 µs ± 24.6 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Mediana inferior dos salario = 3572.581391293322


In [23]:
# usando metodos numpy
%timeit np.percentile(lista_salarios, 25)
print("Mediana inferior dos salario = " + str(np.percentile(lista_salarios, 25)))

119 µs ± 5.59 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Mediana inferior dos salario = 3572.301220300151


### Mediana superior

In [24]:
# usando sem metodos numpy
%timeit salarios.medianaSuperior()
print("Mediana superior dos salario = " + str(salarios.medianaSuperior()))

277 µs ± 28.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Mediana superior dos salario = 8082.598657311342


In [26]:
# usando metodos numpy
%timeit np.percentile(lista_salarios, 75)
print("Mediana superior dos salario = " + str(np.percentile(lista_salarios, 75)))

128 µs ± 6.77 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Mediana superior dos salario = 8074.981613190894


## Medidas de espalhamento

### Desvio padrao populacional

In [29]:
# usando sem metodos numpy
%timeit salarios.desvioPadraoPopulacional()
print("Desvio populacional dos salario = " + str(salarios.desvioPadraoPopulacional()))

421 µs ± 32 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Desvio populacional dos salario = 2536.5054176237713


In [30]:
# usando metodos numpy
%timeit np.std(lista_salarios)
print("Desvio populacional dos salario = " + str(np.std(lista_salarios)))

37.3 µs ± 8.3 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Desvio populacional dos salario = 2536.5054176237722


### Variancia populacional

In [31]:
# usando sem metodos numpy
%timeit salarios.varianciaPopulacional()
print("Variancia populacional dos salario = " + str(salarios.varianciaPopulacional()))

401 µs ± 17.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Variancia populacional dos salario = 6433859.733634743


In [32]:
# usando metodos numpy
%timeit np.var(lista_salarios)
print("Variancia populacional dos salario = " + str(np.var(lista_salarios)))

26 µs ± 1.64 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Variancia populacional dos salario = 6433859.733634747


### Desvio padrao amostral

In [33]:
# usando sem metodos numpy
%timeit salarios.desvioPadraoAmostral()
print("Desvio amostral dos salario = " + str(salarios.desvioPadraoAmostral()))

419 µs ± 41.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Desvio amostral dos salario = 2537.774622315467


In [34]:
# usando metodos numpy
%timeit np.std(lista_salarios, ddof=1)
print("Desvio amostral dos salario = " + str(np.std(lista_salarios, ddof=1)))

29.5 µs ± 3.41 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Desvio amostral dos salario = 2537.7746223154677


### Variancia amostral

In [35]:
# usando sem metodos numpy
%timeit salarios.varianciaAmostral()
print("Variancia amostral dos salario = " + str(salarios.varianciaAmostral()))

424 µs ± 35.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
Variancia amostral dos salario = 6440300.033668411


In [36]:
# usando metodos numpy
%timeit np.var(lista_salarios, ddof=1)
print("Variancia amostral dos salario = " + str(np.var(lista_salarios, ddof=1)))

27 µs ± 1.91 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
Variancia amostral dos salario = 6440300.033668415


## Resultados

> Com execeção da mediana, ao usar numpy os resultados foram obtidos de forma muito mais rapida. Esse resultado em espcifico pode ser devido ao numero pequeno de elementos na lista.
> 
> Abaixo uma tabela com os tempos de execução para uma lista com 1000 elementos.

| Metodo | Tempo Lista | Tempo Numpy |
| ------ | ----------- | ----------- |
| Média aritmética | 103 µs | 7.64 µs |
| Média geométrica | 142 µs | 12.2 µs |
| Média harmonica | 207 µs | 10.9 µs |
| Mediana | 16.4 µs | 33.6 µs |
| Mediana inferior | 269 µs | 119 µs |
| Mediana superior | 277 µs | 128 µs |
| Desvio padrao populacional | 421 µs | 37.3 µs |
| Variancia populacional | 401 µs | 26 µs |
| Desvio padrao amostral | 419 µs | 29.5 µs |
| Variancia amostral | 424 µs | 27 µs  |
